File: leetcode/chain/148.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def merge2List(self, l1: ListNode, l2: ListNode) -> ListNode:
        ret = ListNode()
        last = ret
        while l1 and l2:
            if l1.val <= l2.val:
                last.next = l1
                l1 = l1.next
            else:
                last.next = l2
                l2 = l2.next
            last = last.next
        last.next = l1 if l2 is None else l2
        return ret.next

    def sortList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head

        # 快慢指针求中间位置
        slow = head
        fast = head.next  #!!!很关键
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
        m = slow.next
        slow.next = None  #关键

        l1 = self.sortList(head)
        l2 = self.sortList(m)
        return self.merge2List(l1, l2)

File: leetcode/chain/148.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode* merge_2_list(ListNode* l1, ListNode * l2) {
    ListNode *ret = new ListNode();
    ListNode *last = ret;
    while (l1 && l2) {
      if (l1->val <= l2->val) {
        last->next = l1;
        l1 = l1->next;
      } else {
        last->next = l2;
        l2 = l2->next;
      }
      last = last->next;
    }

    last->next = l1 ? l1 : l2;
    return ret->next;
  }

  ListNode* merge_sort_list(ListNode* node) {
    if (!(node && node->next)) { // nullptr或单个节点 return
      return node;
    }

    // 找中间位置
    ListNode *slow = node;
    ListNode *fast = node->next; //很关键. 如果是node, 那么两个节点时, 当前轮次跟下一轮次参数一样, 不会return
    while (fast && fast->next) {
      slow = slow->next;
      fast = fast->next->next;
    }
    ListNode *second = slow->next;
    slow->next = nullptr; //关键

    ListNode* l1 = merge_sort_list(node);
    ListNode* l2 = merge_sort_list(second);
    return merge_2_list(l1, l2);
  }
};

int main() {
  ListNode* node = new ListNode(3);
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->append(new ListNode(2));
  node->append(new ListNode(4));
  node->append(new ListNode(6));
  node->append(new ListNode(7));
  node->print();

  Solution solution = Solution();
  ListNode* ret = solution.merge_sort_list(node);
  ret->print();
}